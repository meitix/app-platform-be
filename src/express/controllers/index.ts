export * from './payment-plan.controller';
export * from './application.controller';
export * from './auth.controller';
export * from './payment.controller';
// export * from './user.controller';