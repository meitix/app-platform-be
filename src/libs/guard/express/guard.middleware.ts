import { IGuardService } from '..';

/**
 * Returns a middleware that allows access to users who has the required claims (permissions).
 * @param claims required claims to access the route.
 */
export function allow(guardService: IGuardService, claims?: string[]) {
  return (req: any, res: any, next: Function) => {
    if (guardService.hasPermissions(req.user, claims)) {
      next();
    } else {
      res.status(403).send({ message: 'Forbidden' });
    }
  };
}
