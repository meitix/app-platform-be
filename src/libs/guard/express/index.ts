export * from './guard.middleware';
export * from './route-config.interface';
export * from './secure';