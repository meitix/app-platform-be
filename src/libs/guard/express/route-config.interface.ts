export interface IRouteConfig {
    method: HttpMethods,
    path: string;
    claims: string[];
}

export type HttpMethods = 'get' | 'post' | 'put' | 'delete' | 'options' | 'patch' | 'head';