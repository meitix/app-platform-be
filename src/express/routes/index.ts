import { ILogger } from '../../logger';
import { Router } from 'express';
import { iocContainer, TYPES } from '../../ioc';
import { applicationRouter } from './application.routes';
import { ApplicationController, AuthController, RoleController, UserController, PaymentPlansController } from '../controllers';
import { authRouter } from './auth.routes';
import { roleRouter } from './role.routes';
import { userRouter } from './user.routes';
import { paymentPlanRouter } from './payment-plan.routes';

class UserDirectoryRouter {
  router: Router;
  constructor(protected logger: ILogger) {
    this.router = Router();
    this.init();
  }
  init() {
    this.router.use('/applications', applicationRouter(iocContainer.get(ApplicationController), this.logger));
    this.router.use('/auth', authRouter(iocContainer.get(AuthController), this.logger));
    this.router.use('/payment', authRouter(iocContainer.get(AuthController), this.logger));
    this.router.use('/payment-plans', paymentPlanRouter(iocContainer.get(PaymentPlansController), this.logger));
    this.router.use('/role', roleRouter(iocContainer.get(RoleController), this.logger));
    this.router.use('/user', userRouter(iocContainer.get(UserController), this.logger));
  }
}

const userDirectoryRouter = (logger: ILogger) => new UserDirectoryRouter(logger).router;
export { userDirectoryRouter };
