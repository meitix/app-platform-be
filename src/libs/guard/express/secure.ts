import { IRouteConfig } from '.';
import { Application } from 'express';
import { allow } from './guard.middleware';

export function secure(app: Application, config: IRouteConfig[]) {
  config.forEach(item => {

    /**
     * the following line will add the middleware to dynamic path with dynamic method
     * the outcome will be something like this:
     * app.get('/route', (req,res,next) => {});
     */
    app[item.method](item.path, allow(item.claims));
  });
}
