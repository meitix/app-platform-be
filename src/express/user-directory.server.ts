import 'reflect-metadata';
import './controllers';

import * as express from 'express';
import { userDirectoryRouter } from './routes';
import * as swaggerUi from 'swagger-ui-express';
import { DBConfiguration, initializeDatabase, clean } from '../services';
import { ILogger } from '../logger';
import { logMiddleware, createAuthMiddleware } from './middlewares';
import { iocContainer, TYPES } from '../ioc';
import { IIdentityConfig } from '../libs/identity/interfaces';
import { AUTHENTICATED_ROUTES } from './routes/authenticated-routes.const';
import { secure, IRouteConfig } from '../libs/guard/express';
import { seedDB } from '../services/bootstrap/seeder';

export class UserDirectoryServer {
  app: express.Express;
  protected logger: ILogger;

  constructor(protected config: UserDirectoryConfiguration) {
    this.logger = config.logger || (console as any);
    this.app = express();
  }

  private async initialize() {
    this.app.use(express.json());
    this.handleUncaughtExceptions();
    this.secure(AUTHENTICATED_ROUTES);
    this.app.use(userDirectoryRouter(this.logger))
  }

  /**
   * Log all incoming requests and outgoing response.
   */
  public enableRequestLogging(logger?: ILogger) {
    this.app.use(logMiddleware(logger || this.logger));
  }

  /**
   * This method is responsible for cleaning the project and refresh everything such as database and uploaded files.
   * NOT RECOMMENDED TO BE USED ON PRODUCTION ENVIRONMENT.
   */
  public clean() {
    // TODO: Uploaded file cleaner added.
    return clean();
  }

  /**
   * this function will seed database with some default values.
   * @param values default values.
   */
  public seedDatabase(values: any) {
   return seedDB(values);
  }

  public enableDocumentation(url: string) {
    try {
      const swaggerDocument = require('../../swagger.json');
      this.app.use(url, swaggerUi.serve, swaggerUi.setup(swaggerDocument));
    } catch (err) {
      this.logger.error('Unable to load swagger.json', err);
    }
  }

  private handleUncaughtExceptions() {
    process.on('unhandledRejection', this.logger.error.bind(this));
    process.on('uncaughtException', this.logger.error.bind(this));
    process.on('warning', this.logger.warn.bind(this));
  }

  secure(routes: IRouteConfig[]) {
    const authMiddleware = createAuthMiddleware(iocContainer.get<IIdentityConfig>(TYPES.IIdentityConfig).secretKey);
    for (let route of routes) {
      this.app.use(route.path, authMiddleware);
    }
    secure(this.app, routes, iocContainer.get(TYPES.IGuardService));
  }

  public async start(port: string | number) {
    await initializeDatabase(this.config.dbConfig);
    this.initialize();
    this.logger.info(`Connected to database ${this.config.dbConfig.database} on server ${this.config.dbConfig.host}`);
    this.app.listen(port, () => this.logger.info(`User Directory is started on port ${port}`));
  }
}

export interface UserDirectoryConfiguration {
  logger?: ILogger;
  dbConfig: DBConfiguration;
}
