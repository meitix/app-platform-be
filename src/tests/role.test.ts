import 'mocha';
import {userDirectory as UDPromise} from '..';
import * as chai from 'chai';
import chaiHttp = require('chai-http');
import { Express } from 'express';

chai.use(chaiHttp);
let should = chai.should();
const expect = chai.expect;

let sample = {
  name: 'string',
  price: 0,
  dateRange: 0,
  isActive: true,
  description: 'string'
};

let id = 1;

let userDirectory: Express;

before(async () => {
  userDirectory = (await UDPromise).app;
});
