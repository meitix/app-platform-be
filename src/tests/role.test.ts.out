import 'mocha';
import {userDirectory as UDPromise} from '..';
import * as chai from 'chai';
import chaiHttp = require('chai-http');
import { Express } from 'express';
import { iocContainer } from '../ioc';
import { RoleController } from '../express/controllers';
import { Role } from '../libs/guard';
import { APPLICATION_CREATE } from '../config/permissions.const';

chai.use(chaiHttp);
let should = chai.should();
const expect = chai.expect;

let sample = {
  name: 'string',
  price: 0,
  dateRange: 0,
  isActive: true,
  description: 'string'
};


let userDirectory: Express;
let roleController = iocContainer.get(RoleController);

before(async () => {
  userDirectory = (await UDPromise).app;
});

describe('Role Controller', () => {
  it(`Should get user's role with it's children`, async () => {
    const req = { 
    'user.title': 'Super Admin'
    }
    const result = await roleController.fetch(req as any);
    result.should.be.instanceof(Array);
  });
});

describe('Should create new role', async () => {
  const role = new Role('test', [APPLICATION_CREATE]);
  const result = await roleController.create(role);
  expect(result).equal(role);
});
